src/condition/ssrf_detector.cpp

```diff
@@ -0,0 +1,279 @@
+{
+    const auto path_end = uri.path_index + uri.path.size();
+
+    // Check if the application is giving full control of the path to the user
+    // either fully or after a forward-slash.
+    if (!uri.path.empty()) {
+        //  scheme://userinfo@host:port/path?query#fragment
+        //                             ─────
+        if (uri.path.size() == param.size()) {
+            return false;
+        }
+        //  scheme://userinfo@host:port/path?query#fragment
+        //                              ────
+        if (uri.path[0] == '/' && param_index == uri.path_index + 1 &&
+            param.size() == uri.path.size() - 1) {
+            return false;
+        }
+    }
+
+    // REST parameter injection will involve introducing a / to the URL, however, such
+    // slashes are allowed after the ? or the #. Check if the slash is within the path
+    //
+    //  scheme://userinfo@host:port/path?query#fragment
+    //                             ─────
+    if (auto slash_index = param.find('/'); slash_index != npos) {
+        slash_index += param_index;
+        if (slash_index < path_end) {
+            // The path is partially under control of the user, this might be intentional
+            // so lets check for a possible LFI
+            auto relative_dir_index = param.find("..");
+            while (relative_dir_index != npos) {
+
+                // We found '..', check if it's enclosed by '/'
+                auto dir_index = relative_dir_index + param_index;
+                if (dir_index < path_end && (dir_index + 2) < uri.raw.size() &&
+                    uri.raw[dir_index - 1] == '/' && uri.raw[dir_index + 2] == '/') {
+                    return true;
+                }
+                relative_dir_index = param.find("..", relative_dir_index + 2);
+            }
+        }
+    }
+
+    // The application giving total control to the user after a given point.
+    // If it starts in the path, assume it's intentional
+    //
+    //  scheme://userinfo@host:port/path?query#fragment
+    //                             ────────────────────
+    if ((uri.query.empty() && uri.fragment.empty()) ||
+        (param_index < path_end && (param_index + param.size()) == uri.raw.size())) {
+        return false;
+    }
+
+    // Check if the entire query string was injected
+    //
+    //  scheme://userinfo@host:port/path?query#fragment
+    //                                 <─────
+    const auto query_index = param.find('?');
+    if (uri.query_index != npos && query_index != npos &&
+        (param_index + query_index + 1) == uri.query_index) {
+        // We had some cases where this was expected behavior, to make sure
+        // we check whether the next character is a '&'
+        auto after_param = uri.raw.substr(param_index + param.length());
+        if (!after_param.empty() && after_param[0] == '&') {
+            return false;
+        }
+
+        // We can also check if the there's a valid parameter in key=value form
```

Comment:
Don't get it. What's the rationale here? How did the user format the uri, and what did he expect the param to look like?

Reply:
There was a test case in the original code covering this particular condition:

```
url: http://core-goals/v1/projects/42/goals?projectId=42&
input:  /v1/projects/42/goals?
```

Reply:
I can understand the case where you have

https://example.com/foo/{param}&a=b
with param sth like bar?c=d

or https://example.com/foo/{param}

or even (more questionable)
https://example.com/foo/{param}c=d
with param = bar? or param = bar?a=b&

but this matches even if we have param = bar?a=b

the result is something that is almost certainly a injection: they key-value pair (a, bc=d).

Reply:
I'm leaning towards removing this special case as I can't think of a case that isn't questionable...

Reply:
I finally decided to remove the check, I think the existing checks should be enough, but we can reevaluate later.